### PR TITLE
ARTEMIS-3285 potential duplicate messages with LVQ + non-destructive

### PR DESCRIPTION
**Pull request by jbertram**

### CODE REVIEW

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java [hunk 1]
```diff
@@ -588,4 +594,82 @@ private void sendLVQTombstone(ConnectionSupplier producerConnectionSupplier, Str
          producer.send(message1, javax.jms.Message.DEFAULT_DELIVERY_MODE, javax.jms.Message.DEFAULT_PRIORITY, tombstoneTimeToLive);
       }
    }
+
+   @Test
+   public void testMultipleLastValuesCore() throws Exception {
+      testMultipleLastValues(CoreConnection);
```
Comment (clebertsuconic) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java:600]: Can you instead use CFUtil.createConnectionFactory("AMQP", "tcp://localhost:61616")

or 

CFUtil.createConnectionFactory("CORE", "tcp://localhost:61616")


We should probably remove the Connection supplier. It's a bit simpler to read.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java [hunk 2]
```diff
@@ -588,4 +594,82 @@ private void sendLVQTombstone(ConnectionSupplier producerConnectionSupplier, Str
...
+         }
+      }
+      for (Map.Entry<String, List<String>> entry : results.entrySet()) {
+         System.out.print("Messages received with lastval=" + entry.getKey() + " (");
+         for (String s : entry.getValue()) {
+            int occurrences = Collections.frequency(entry.getValue(), s);
+            if (occurrences > 1 && !dups.containsValue(Integer.parseInt(s))) {
+               dups.put(s, occurrences);
+            }
+            System.out.print(s + ",");
```
Comment (clebertsuconic) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java:661]: Can you use log.info

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java [hunk 3]
```diff
@@ -588,4 +594,82 @@ private void sendLVQTombstone(ConnectionSupplier producerConnectionSupplier, Str
...
+         });
+
+         t.start();
+         int received = 0;
+         while (true) {
+            TextMessage tm = (TextMessage) consumer.receive(500);
+            if (tm == null) {
+               break;
+            }
+            System.out.println("Received: " + tm.getStringProperty("data") + "; total: " + ++received);
```
[the remark this answers is not in the document]
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java]: Done.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java (hunk 2 above)
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java]: Done.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java [hunk 4]
```diff
@@ -350,25 +342,33 @@ public SimpleString getLastValueKey() {
...
+         return true;
+      }
+   }
+
    private static class HolderReference implements MessageReference {
 
       private final SimpleString prop;
 
-      private volatile boolean delivered = false;
+      private volatile AtomicBoolean delivered = new AtomicBoolean(false);
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java:358]: Please use the AtomicLongFieldUpdater instead?

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java [hunk 5]
```diff
@@ -588,4 +597,82 @@ private void sendLVQTombstone(ConnectionSupplier producerConnectionSupplier, Str
...
+
+      }
+      if (dups.size() > 0) {
+         StringBuffer sb = new StringBuffer();
+         for (Map.Entry<String, Integer> stringIntegerEntry : dups.entrySet()) {
+            sb.append(stringIntegerEntry.getKey() + "(" + stringIntegerEntry.getValue() + "),");
+         }
+         Assert.fail("Duplicate messages received " + sb);
+      }
+
```
Comment (clebertsuconic) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/JMSNonDestructiveTest.java:676]: I added the following code here:

```java
      HashMap<Integer, AtomicInteger> map = countJournal(server.getConfiguration());

      map.forEach((a, b) -> System.out.println("record type " + a + ", counts=" + b));

```

That would be all you need to check on the journal...


Right now the output here shows exactly what I predicted:


```
record type 32, counts=100
record type 33, counts=97
record type 34, counts=3
record type 45, counts=100
```

You will need some assertion on those records. You need to ack the previous message that was already delivered.


If you can do that on handle(MessageReference), make it specialized on LastValueQueue would be better than the method you added on the consumer.

**Comment by clebertsuconic** [comment]
there are also a few failures added on this PR:

org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveCoreProducerCoreConsumer[persistenceEnabled=false, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveAMQPProducerAMQPConsumer[persistenceEnabled=false, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveLVQWithConsumerFirstAMQP[persistenceEnabled=false, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveLVQWithConsumerFirstCore[persistenceEnabled=false, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveCoreProducerAMQPConsumer[persistenceEnabled=false, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveAMQPProducerCoreConsumer[persistenceEnabled=false, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveCoreProducerCoreConsumer[persistenceEnabled=true, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveAMQPProducerAMQPConsumer[persistenceEnabled=true, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveLVQWithConsumerFirstAMQP[persistenceEnabled=true, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveLVQWithConsumerFirstCore[persistenceEnabled=true, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveCoreProducerAMQPConsumer[persistenceEnabled=true, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveAMQPProducerCoreConsumer[persistenceEnabled=true, scanPeriod=100]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveCoreProducerCoreConsumer[persistenceEnabled=true, scanPeriod=-1]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveAMQPProducerAMQPConsumer[persistenceEnabled=true, scanPeriod=-1]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveLVQWithConsumerFirstAMQP[persistenceEnabled=true, scanPeriod=-1]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveLVQWithConsumerFirstCore[persistenceEnabled=true, scanPeriod=-1]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveCoreProducerAMQPConsumer[persistenceEnabled=true, scanPeriod=-1]
org.apache.activemq.artemis.tests.integration.amqp.JMSNonDestructiveTest.testNonDestructiveAMQPProducerCoreConsumer[persistenceEnabled=true, scanPeriod=-1]
org.apache.activemq.artemis.tests.integration.cli.RecoverTest.org.apache.activemq.artemis.tests.integration.cli.RecoverTest
org.apache.activemq.artemis.tests.integration.cluster.failover.ReplicatedMultipleServerFailoverExtraBackupsTest.testStartLiveFirst
org.apache.activemq.artemis.tests.integration.jms.client.LVQTest.testLVQandNonDestructive

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java (hunk 4 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java:358]: As far as I can tell you can't use `AtomicLongFieldUpdater` on a `boolean`.

**Comment by jbertram** [comment]
I'll go back to the drawing board and rework this.

**Pull request closed without merging**